src/NServiceBus.Core/EndpointConfiguration.cs

```diff
@@ -218,7 +218,9 @@ internal InitializableEndpoint Build()
                 Settings.SetDefault("PublicReturnAddress", publicReturnAddress);
             }
 
-            Settings.SetDefault<Conventions>(conventionsBuilder.Conventions);
+            var conventions = conventionsBuilder.Conventions;
+            Settings.SetDefault<Conventions>(conventions);
+            Settings.SetDefault("MessageTypes", new Lazy<Type[]>(() => Settings.GetAvailableTypes().Where(t => conventions.IsMessageType(t)).ToArray()));
```

Comment:
Good question. I was not sure in which context it is going to be invoked, Since the extension is on ReadOnlySettings which should only be called in feature I figured it is best to compute once in a lazy fashion (with that way it is also thread safe in case someone calls it from weird locations).